### DISCUSSION TITLE
Customizing GraphQL paging and sorting

### DISCUSSION THREAD

**Discussion by BringerOD** [ghiscoding/aurelia-slickgrid#703] [Ideas]
### Your Environment
| Software          | Version(s) |
| ----------------- | ---------- |
| Aurelia           |  1.0
| Aurelia-Slickgrid |  4.1.1
| TypeScript        | 4.0.2
| Operating System  | Win10
| NPM/Node/Yarn     | YARN

### Context
I am using [Hot Chocolate 12.5](https://chillicream.com/) and they have a different syntax for sorting, paging and I assume filtering.  I have not got to filtering with the grid yet.  Hot chocolate is getting popular and is easy to use.

Here is a sample query that has skip and take instead of offset and limit.  I was able to get replace these with little effort.  Also, was able to make a new object from the result to change the return type to nodes instead of items.

`query {
  connections(
    skip: 0
    take: 50
    order: {  createdDate:  ASC, user: {name:  DESC}}
  ) {
    totalCount
    items {
     createdDate
     user {
        name
      }
    }
  }
}
`

They use order instead of orderby, which will be easy to replace as well.   The rest is a little harder, with the differences in complex object sorting and the syntax in general.

Here is an example of what I see being requested by the grid
`
 query   {  connections (offset:0,first:50,orderBy:[{field:phoneNumber,direction:ASC}])  { totalCount, nodes   { id,createdDate,phoneNumber,to,from } } }
`

### Possible Solution
Is there a way to override or extend the current implementation of the GraphqlService?

Are there any plans to support the default syntax from Hot Chocolate?

Thank you for working on this repo.

**Comment by ghiscoding** [comment]
It won't work out of the box and when we started our project we had to define a structure and at the time there wasn't any standard for querying with pagination and stuff, so we decided to go with what made sense for us and was a good foundation for everyone to use. Since there wasn't any standard, it won't be a direct match with external GraphQL lib like the one you mentioned (first time I hear about it) and the only possible way would be to add a parser of some kind. 

The easiest is probably to extend the service like you mentioned, I revisited all my services at some point and made them all as `protected` instead of `private` to make it possible to extend. The GraphQL Service you're looking for, is now in [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) monorepo ([GraphqlService](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/graphql/src/services/graphql.service.ts)) and I also use this external lib [GraphQL Query Builder](https://github.com/codemeasandwich/graphql-query-builder) to help creating a valid query string (it however didn't have TypeScript support, so I rewrote it all in TS since it's small anyway and is now located in my lib here [graphqlQueryBuilder.ts](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/graphql/src/services/graphqlQueryBuilder.ts)). I have created 2 backend services over the years (OData and GraphQL) and they both follow these 3 interfaces to work properly with SlickGrid
- [backendService.interface](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/interfaces/backendService.interface.ts)
- [backendServiceApi.interface](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/interfaces/backendServiceApi.interface.ts)
- [backendServiceOption.interface](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/interfaces/backendServiceOption.interface.ts)

So you probably want to extend the service, you could also eventually choose to add your port into [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) as another optional package without impacting other users (that's the beauty of a monorepo).

**Comment by BringerOD** [comment]
Thanks for the quick response!

I will give it a shot.  If I come up with something that others can use, I will submit it. 

Is there a standard now?  I did not see one.

**Comment by ghiscoding** [comment]
I'll turn this as a Discussion since it's not really an issue and is better to discuss

**Comment by BringerOD** [comment]
I did start the project and implemented all the interfaces into new classes, copied the implementation from your work.  Was able to make some minor changes successfully, then hit a wall on the order by and filters implementation. 

I mainly work on backend systems, and I realized my typescript skills need to be updated and it would take me more time than I have to do it.  I have tabled it for now and will have another team member look at it in the coming weeks.
